### PR TITLE
test: run CSI driver e2e tests on Windows clusters

### PR DESCRIPTION
**Pull request by chewong**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

This PR allows future PRs to build and push Azure Disk CSI Driver Windows containers to ACR using Travis CI. This will be the first step in enabling e2e tests on Windows clusters. The Windows PR jobs should be triggered assuming that the Windows image has been pushed to ACR.

/assign @andyzhangx 
/cc @ZeroMagic 

Ref: https://github.com/ZeroMagic/azuredisk-csi-driver/commit/b3848134fd90bcbdf50284eb0ff5a2c79b81420f

**Which issue(s) this PR fixes**:

https://github.com/kubernetes-sigs/azuredisk-csi-driver/issues/257

**Special notes for your reviewer**:

**Release note**:
```

```

### CODE REVIEW

**Comment by chewong** [comment]
/test pull-azuredisk-csi-driver-e2e-single-az
--

File: .travis.yml [hunk 1]
```diff
@@ -5,9 +5,20 @@ dist: bionic
...
 script:
-  - go test -covermode=count -coverprofile=profile.cov ./pkg/...
-  - $GOPATH/bin/goveralls -coverprofile=profile.cov -service=travis-ci
+  - |
+    if [[ "${TRAVIS_OS_NAME}" == "linux" ]]; then
+      go test -covermode=count -coverprofile=profile.cov ./pkg/...
+      GO111MODULE=off go get github.com/mattn/goveralls
+      $GOPATH/bin/goveralls -coverprofile=profile.cov -service=travis-ci
+    elif [[ "${TRAVIS_OS_NAME}" == "windows" ]]; then
+      echo "$PROW_AAD_CLIENT_SECRET" | docker login -u "$PROW_AAD_CLIENT_ID" --password-stdin "$REGISTRY"
```
Comment (ZeroMagic) [inline .travis.yml]: The environment variables are not set in travis-ci. So login failed here. 

Maybe there are some problems with this azure sp. We need to discuss them with Andy @andyzhangx .

File: Makefile [hunk 2]
```diff
@@ -21,6 +21,9 @@ IMAGE_VERSION ?= v0.7.0
 # Use a custom version for E2E tests if we are in Prow
 ifdef AZURE_CREDENTIALS
 override IMAGE_VERSION := e2e-$(GIT_COMMIT)
+ifdef TEST_WINDOWS
```
Comment (ZeroMagic) [inline Makefile]: It seems that this judgment does not take effect in ci. I think it should be `ifdef $TEST_WINDOWS`.

File: .travis.yml (hunk 1 above)
Comment (andyzhangx) [reply .travis.yml]: We cannot store sp in travis-ci any more, it would leak secret.

**Comment by chewong** [comment]
I think we still have permission to store SP in Travis CI (I just added one before opening this PR). If you click "Details" below, "More Options", and "Settings". To not leak secret, we just need to disable "DISPLAY VALUE IN BUILD LOG" when declaring the environment variables.

image

**Comment by chewong** [comment]
"Encrypted variables are not available to untrusted builds such as pull requests coming from another repository." - https://docs.travis-ci.com/user/environment-variables/#defining-encrypted-variables-in-travisyml. That's why the registry credential isn't available to the pull request.

**Comment by chewong** [comment]
We can actually use [`az acr build`](https://docs.microsoft.com/en-us/cli/azure/acr?view=azure-cli-latest#az-acr-build) to directly build and push the image to ACR in Prow.

```
$ az acr build --registry chuwon -t chuwon.azurecr.io/azuredisk-csi:test -f ./pkg/azurediskplugin/Windows.Dockerfile --platform windows .
Packing source code into tar to upload...
Excluding '.gitignore' based on default ignore rules
Excluding '.git' based on default ignore rules
Uploading archived source code from '/var/folders/8f/tc_97tds47d9c3ckr6d1kxsm0000gn/T/build_archive_f672d4dc370c41b2b8da1090207b0699.tar.gz'...
Sending context (49.793 MiB) to registry: chuwon...
Queued a build with ID: cc1
Waiting for an agent...
2020/03/27 16:01:34 Downloading source code...
2020/03/27 16:01:51 Finished downloading source code
2020/03/27 16:01:54 Using acb_vol_66ad55ea-4da8-4ff4-9cfb-cce6a9500afd as the home volume
2020/03/27 16:01:54 Setting up Docker configuration...
2020/03/27 16:02:05 Successfully set up Docker configuration
2020/03/27 16:02:05 Logging in to registry: chuwon.azurecr.io
2020/03/27 16:02:10 Successfully logged into chuwon.azurecr.io
2020/03/27 16:02:10 Executing step ID: build. Timeout(sec): 28800, Working directory: '', Network: ''
2020/03/27 16:02:10 Scanning for dependencies...
2020/03/27 16:02:15 Successfully scanned dependencies
2020/03/27 16:02:15 Launching container with name: build
Sending build context to Docker daemon  171.6MB
Step 1/7 : FROM mcr.microsoft.com/windows/servercore:1809 as core
1809: Pulling from windows/servercore
65014b3c3121: Already exists
23645bd7ee08: Already exists
Digest: sha256:9d0dff5e064c87d612503e1559c3d14353b404a1cd4b53ac1373e0fc62017971
Status: Downloaded newer image for mcr.microsoft.com/windows/servercore:1809
 ---> 73d9a9b9c9fb
Step 2/7 : FROM mcr.microsoft.com/windows/nanoserver:1809
1809: Pulling from windows/nanoserver
Digest: sha256:5de6bd32bd453d60c8f549d28845552e89ad3652566e141ac82023b6ba10374d
Status: Image is up to date for mcr.microsoft.com/windows/nanoserver:1809
 ---> 39a578dbf5b5
Step 3/7 : LABEL description="CSI Azure disk plugin"
 ---> Running in 9749c814f0a3
Removing intermediate container 9749c814f0a3
 ---> 7a87bd8268d7
Step 4/7 : COPY ./_output/azurediskplugin.exe /azurediskplugin.exe
 ---> 490910a42856
Step 5/7 : COPY --from=core /Windows/System32/netapi32.dll /Windows/System32/netapi32.dll
 ---> 43e7f767bae5
Step 6/7 : USER ContainerAdministrator
 ---> Running in 62800d67e8e1
Removing intermediate container 62800d67e8e1
 ---> 2ee74d0039bf
Step 7/7 : ENTRYPOINT ["/azurediskplugin.exe"]
 ---> Running in ba47b0eeaf5f
Removing intermediate container ba47b0eeaf5f
 ---> 7b9612b2460c
Successfully built 7b9612b2460c
Successfully tagged chuwon.azurecr.io/azuredisk-csi:test
2020/03/27 16:02:39 Successfully executed container: build
2020/03/27 16:02:39 Executing step ID: push. Timeout(sec): 1800, Working directory: '', Network: ''
2020/03/27 16:02:39 Pushing image: chuwon.azurecr.io/azuredisk-csi:test, attempt 1
The push refers to repository [chuwon.azurecr.io/azuredisk-csi]
9f7dc47abcac: Preparing
18483ae27522: Preparing
4566caffc629: Preparing
de981866e149: Preparing
f4e46717ece7: Preparing
cea17d6bca65: Preparing
cea17d6bca65: Waiting
18483ae27522: Pushed
4566caffc629: Pushed
9f7dc47abcac: Pushed
f4e46717ece7: Pushed
cea17d6bca65: Skipped foreign layer
de981866e149: Pushed
test: digest: sha256:fc495593b9ed29e5e2241023361d46e3e51deb3732cec0943ec65b36f4057202 size: 1751
2020/03/27 16:02:47 Successfully pushed image: chuwon.azurecr.io/azuredisk-csi:test
2020/03/27 16:02:47 Step ID: build marked as successful (elapsed time in seconds: 28.690174)
2020/03/27 16:02:47 Populating digests for step ID: build...

2020/03/27 16:03:03 Successfully populated digests for step ID: build
2020/03/27 16:03:03 Step ID: push marked as successful (elapsed time in seconds: 8.321045)
2020/03/27 16:03:03 The following dependencies were found:
2020/03/27 16:03:03
- image:
    registry: chuwon.azurecr.io
    repository: azuredisk-csi
    tag: test
    digest: sha256:fc495593b9ed29e5e2241023361d46e3e51deb3732cec0943ec65b36f4057202
  runtime-dependency:
    registry: mcr.microsoft.com
    repository: windows/nanoserver
    tag: "1809"
    digest: sha256:5de6bd32bd453d60c8f549d28845552e89ad3652566e141ac82023b6ba10374d
  buildtime-dependency:
  - registry: mcr.microsoft.com
    repository: windows/servercore
    tag: "1809"
    digest: sha256:9d0dff5e064c87d612503e1559c3d14353b404a1cd4b53ac1373e0fc62017971
  git: {}


Run ID: cc1 was successful after 1m31s
```

**Comment by andyzhangx** [comment]
@chewong ACR build pipeline is great, share you the scripts how I build linux/windows images now: https://github.com/kubernetes-sigs/azuredisk-csi-driver/tree/master/pkg/azurediskplugin

**Comment by chewong** [comment]
Pending PRs to make this PR work: https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/340, https://github.com/kubernetes/test-infra/pull/17020

File: Makefile [hunk 3]
```diff
@@ -92,21 +104,31 @@ azuredisk-windows:
...
 	CGO_ENABLED=0 GOOS=windows go build -a -ldflags ${LDFLAGS} -o _output/azurediskplugin.exe ./pkg/azurediskplugin
 
-.PHONY: container
-container: azuredisk
-	docker build --no-cache -t $(IMAGE_TAG) -f ./pkg/azurediskplugin/Dockerfile .
-
 .PHONY: azuredisk-container
-azuredisk-container: azuredisk
+azuredisk-container:
+# ifdef CI
```
Comment (andyzhangx) [inline Makefile:109]: why comment here?
The two dependencies have been merged: https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/340, https://github.com/kubernetes/test-infra/pull/17020

Comment (chewong) [reply Makefile:109]: I was making sure that the CI for Linux cluster pass. I will uncomment it after the new test image reflects kubernetes/test-infra#17020.

File: Makefile [hunk 4]
```diff
@@ -66,13 +68,19 @@ e2e-test:
 
 .PHONY: e2e-bootstrap
 e2e-bootstrap: install-helm
-	# Only build and push the image if it does not exist in the registry
 	docker pull $(IMAGE_TAG) || make azuredisk-container push
+ifdef TEST_WINDOWS
+	helm install azuredisk-csi-driver charts/latest/azuredisk-csi-driver --namespace kube-system --wait --timeout=15m -v=5 --debug \
+		--set image.azuredisk.repository=$(REGISTRY)/$(IMAGE_NAME) \
+		--set image.azuredisk.tag=$(IMAGE_VERSION) \
+		--set controller.replicas=1 \
```
Comment (andyzhangx) [inline Makefile:76]: remove this `--set controller.replicas=1` since we already have 3 master node now, let's use the default value(2)

**Comment by chewong** [comment]
E2E test works on my local cluster. @ZeroMagic could you help change the file structure of https://k8scsi.blob.core.windows.net/csi-proxy/v0.1.0/binaries/csi-proxy.tar.gz to 
```
.
├── bin
│   └── csi-proxy.exe
└── build
    └── server.exe
```

**Comment by ZeroMagic** [comment]
changed it. Please try it again. @chewong

**Comment by chewong** [comment]
Kept encounter the following error in pull-azuredisk-csi-driver-e2e-windows unless I set `controller.replicas` to 1:
```
Error: etcdserver: request timed out
helm.go:75: [debug] etcdserver: request timed out
```

**Comment by andyzhangx** [comment]
I find only `csi-provisioner` is crashing in e2e windows test, could you take a look @ZeroMagic ?
```
  csi-provisioner:
    Container ID:  docker://112fd09b9d62871f7fa80e1fc301e61e3412f3219470006a63a4b8d9352b0401
    Image:         mcr.microsoft.com/oss/kubernetes-csi/csi-provisioner:v1.5.0
    Image ID:      docker-pullable://mcr.microsoft.com/oss/kubernetes-csi/csi-provisioner@sha256:21fae82e312b135c9b09f75041922a4e189ce6d86d23c83a7e1be48dbe172a74
    Port:          <none>
    Host Port:     <none>
    Args:
      --provisioner=disk.csi.azure.com
      --feature-gates=Topology=true
      --csi-address=$(ADDRESS)
      --connection-timeout=15s
      --v=5
      --timeout=120s
      --enable-leader-election
      --leader-election-type=leases
    State:          Running
      Started:      Fri, 03 Apr 2020 02:25:41 +0000
    Last State:     Terminated
      Reason:       Error
      Exit Code:    255
      Started:      Fri, 03 Apr 2020 02:24:20 +0000
      Finished:     Fri, 03 Apr 2020 02:24:50 +0000
    Ready:          True
    Restart Count:  4
    Limits:
      cpu:     200m
      memory:  200Mi
    Requests:
      cpu:     10m
      memory:  20Mi
    Environment:
      ADDRESS:  /csi/csi.sock
    Mounts:
      /csi from socket-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from csi-azuredisk-controller-sa-token-ldr4t (ro)
```

**Comment by chewong** [comment]
From pull-azuredisk-csi-driver-e2e-windows:
```
Apr  3 03:57:38.022: INFO: Deleting pod "azuredisk-volume-tester-v5qfz-675857794b-z79xf" in namespace "azuredisk-8452"
Apr  3 03:57:38.099: INFO: Waiting for pod "azuredisk-volume-tester-v5qfz-675857794b-z79xf" in namespace "azuredisk-8452" to be fully deleted
STEP: checking again that the pod is running
STEP: checking pod exec
Apr  3 03:57:54.394: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/root/tmp072334695/kubeconfig/kubeconfig.westus2.json exec azuredisk-volume-tester-v5qfz-675857794b-jj5vp --namespace=azuredisk-8452 -- powershell.exe cat C:\mnt\test-1\data.txt'
Apr  3 03:57:58.121: INFO: stderr: ""
Apr  3 03:57:58.121: INFO: stdout: "hello world\r\n"
Apr  3 03:58:00.122: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/root/tmp072334695/kubeconfig/kubeconfig.westus2.json exec azuredisk-volume-tester-v5qfz-675857794b-jj5vp --namespace=azuredisk-8452 -- powershell.exe cat C:\mnt\test-1\data.txt'
Apr  3 03:58:06.298: INFO: stderr: ""
Apr  3 03:58:06.298: INFO: stdout: "hello world\r\n"
Apr  3 03:58:08.298: INFO: Unexpected error occurred: Failed to find "hello world
hello world
", last result: "hello world
"
```

Looks like we will need to merge https://github.com/kubernetes-sigs/azuredisk-csi-driver/pull/345 first.

**Comment by ZeroMagic** [comment]
All tests passed. Failed due to the following info:
```
Uninstalling log-dump-daemonset.yaml via kubectl
Error from server: error when deleting "/root/tmp625673963/log-dump-daemonset.yaml": etcdserver: request timed out
```

/test pull-azuredisk-csi-driver-e2e-windows

File: Makefile [hunk 5]
```diff
@@ -92,21 +99,32 @@ azuredisk-windows:
...
+	az acr build --registry $(REGISTRY_NAME) -t $(IMAGE_TAG)-linux-amd64 -f ./pkg/azurediskplugin/Dockerfile --platform linux .
+	az acr build --registry $(REGISTRY_NAME) -t $(IMAGE_TAG)-windows-1809-amd64 -f ./pkg/azurediskplugin/Windows.Dockerfile --platform windows .
+	docker manifest create $(IMAGE_TAG) $(IMAGE_TAG)-linux-amd64 $(IMAGE_TAG)-windows-1809-amd64
+	docker manifest inspect $(IMAGE_TAG)
+else
+ifdef TEST_WINDOWS
+	make azuredisk-windows
+	docker build --no-cache -t $(IMAGE_TAG) -f ./pkg/azurediskplugin/Windows.Dockerfile .
+else
+	make azurediskfi
```
Comment (andyzhangx) [inline Makefile:116]: what is `azurediskfi`? typo?

Comment (chewong) [reply Makefile:116]: It's a typo. Let me open another PR to fix that.

**Pull request merged**